pkg/chunk/storage/factory.go

```diff
@@ -28,12 +28,23 @@ const (
 	StorageEngineTSDB   = "tsdb"
 )
 
+type indexStoreFactories struct {
+	indexClientFactoryFunc IndexClientFactoryFunc
+	tableClientFactoryFunc TableClientFactoryFunc
+}
+
+// IndexClientFactoryFunc defines signature of function which creates chunk.IndexClient for managing index in index store
 type IndexClientFactoryFunc func() (chunk.IndexClient, error)
 
-var customIndexClients = map[string]IndexClientFactoryFunc{}
+// TableClientFactoryFunc defines signature of function which creates chunk.TableClient for managing tables in index store
+type TableClientFactoryFunc func() (chunk.TableClient, error)
 
-func RegisterIndexClient(name string, factory IndexClientFactoryFunc) {
-	customIndexClients[name] = factory
+var customIndexStores = map[string]indexStoreFactories{}
+
+// RegisterIndexStore is used for registering a custom index type.
+// When an index type is registered here with same name as existing types, the registered one takes the precedence.
+func RegisterIndexStore(name string, indexClientFactory IndexClientFactoryFunc, tableClientFactory TableClientFactoryFunc) {
+	customIndexStores[name] = indexStoreFactories{indexClientFactory, tableClientFactory}
```

Comment:
Should we panic, if both functions are nil? (Alternatively, we can have two separate Register functions, each for one type, and panic if supplied function is nil).

Reply:
Since `RegisterIndexStore` is not used anywhere (except in the tests), it's hard to see if this is a problem or not, but in general, there is synchronization missing on access to `customIndexStores` map. It may be worth creating a small struct to keep this map (or two maps), and use a mutex to protect access to it.

Reply:
> Should we panic, if both functions are nil? 

I don't think so. Passing `nil` could be a legit way to unregister an index store, unless we offer an `UnregisterIndexStore` function.

Reply:
> Should we panic, if both functions are nil?
It is up to the user of this to decide whether they need both IndexClient or TableClient or one of them to be custom. Most of the cases should have both of them custom-defined but it would not be the case always. I do not see any value in keeping two different maps and functions to register them separately. It might be just me though.

> It may be worth creating a small struct to keep this map (or two maps), and use a mutex to protect access to it.
I do not think reads and writes would ever be simultaneous. `NewStore` call in Cortex initializes all stores at once and it is expected to have all the custom clients registered by then which means all the writes would be done before `NewStore` is called. Also, the custom client registration and `NewStore` would never run async since they both run in the same goroutine most likely main. I do not see a usecase of ever running them async.
What do you think?

Reply:
> I do not think reads and writes would ever be simultaneous.

It's not about simultaneous reads/writes, but about reading and writing from different goroutines (at different times). To make sure that writes are visible to readers, you need *some* kind of happens-before relationship. Mutex is one possibility. Creating all reading goroutines only after registering these functions is another one. Since it's not used in real code currently, it's not wrong (yet), but lack of synchronization is a red flag.

> NewStore call in Cortex initializes all stores at once and it is expected to have all the custom clients registered by then which means all the writes would be done before NewStore is called.

This would indeed establish happens-before relationship.